z390test/README.md

```diff
@@ -0,0 +1,243 @@
+
+    ./build.sh
+    win> BUILD.BAT
+
+## Run 
+
+By default the tests will not show the output of the `printOutput` method (see below).
+While developing tests, it is useful to see this output.
+
+    export Z390_PRINT_OUTPUT 1
+    win> SET Z390_PRINT_OUTPUT=1
+
+Run the tests from z390test clone directory
+
+    cd z390test
+    ./gradlew test
+    win> gradlew test
+
+If you use an IDE like IntelliJ or Eclipse, they have support for Gradle test suites.
+
+https://www.jetbrains.com/help/idea/work-with-tests-in-gradle.html
+
+
+    
+## Writing test cases
+
+The following is a basic test case that executes an assembly of a module in the z390 codebase 
+and checks the return code.
+
+```groovy
+import org.junit.jupiter.api.Test
+
+class TestZ390Test extends z390Test {
+
+    var options = ['trace', 'noloadhigh', "SYSMAC(${basePath("mac")})"]
+
+    @Test
+    void test_file_source() {
+        int rc = this.asm(basePath('tests', 'TESTINS1'), *options)
+        this.printOutput()
+        assert rc == 0
+    }
+}
+```
+
+To use the standard framework, your class should extend `z390Test`
+which provides methods for running and capturing the output.
+
+The tests use standard JUnit test structures.
+
+## Methods for testing
+
+The z390Test class embeds a number of methods for interacting with z390 similar
+to the scripts you would use.
+
+### assemble from file
+
+```groovy
+int rc = this.asm(basePath('tests', 'TESTINS1'), *options)
+assert rc == 0
+```
+Parameters:
+* asmFile - The HLASM file without prefix to assemble. Use basePath method to build path relative to z390.
+* args... - 0-n parmeters to pass to the asm program, generally options.
+
+Returns:
+* rc - The result code from the execution
+
+Methods available :
+* asm - Assemble only
+* asml - Assemble and link. Any non zero RC on assemble will exit with RC
+* asmlg - Assemble, link and go. Any non zero RC on assemble and link will exit with RC
+* mz390 - Assemble only
+* cblc - COBOL assemble only
+* cblclg - COBOL assemble, link and go. Any non-zero return will exist with RC
+* lz390 - Link only. Output not cleared prior to run
+* ez390 - Run only. Output not cleared prior to run
+
+### assemble from inline source
+
+```groovy
+var source = """TESTB    START 0
+     USING *,13
+     STM   14,12,12(13)
+     ST    13,8(13)
+     ST    15,4(15)
+     LR    13,15
+     J
+     RETURN (14,12)
+     END   TESTB
+"""
+String sourceFilename = createTempFile("INLINE.MLC", source)
+int rc = this.asm(sourceFilename, *options)
+assert rc == 0
+```
+Use the `createTempFile` method to create a file that can be
+used in the assembly.
+
+### Set environment variables
+
+```groovy
+this.env = ['SNAPOUT': basePath('zopcheck', 'SNAPOUT.TXT')]
+```
+The env property can be set with a hashlist of environment variable that will be passed
+to the z390 calls.
+
+To just add a single envvar without replacing the existing:
+```groovy
+this.env.put('SNAPOUT', basePath('zopcheck', 'SNAPOUT.TXT'))
+```
```

Comment:
Without replacing the existing ... what? If SNAPOUT already exists, will this env.put replace it with the new value?

Reply:
Will change to "To just add a single envvar without replacing any previously set environment variables:"
And yes, if SNAPOUT already exists, this command will replace it.. 
this.env is a variable.. If you write `this.env = something` then you are fully replacing it.
If you write `this.env.put('key', 'value')` then you are adding if it does not exist and updating if it does. This is just behavior of the Map, not the framework.